Calculate pool desirability according to delegation design spec

Anviking opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1250 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] New completely isolated module implementing ~`nonMyopicMemberRewards`~ `desirability` and related helper functions from the [spec](https://hydra.iohk.io/build/1389333/download/1/delegation_design_spec.pdf).
- [x] Some simple unit/property tests as sanity checks

# Comments

- To be squashed before merging. Leaving a couple un-squashed for now, for history.

⬇️ Now reverted. Coming in a later PR! (Task #1276 )

Work in progress example ranking (without averaging, and with some hard-coded slightly off values)
<img width="1072" alt="Skärmavbild 2020-01-10 kl  18 46 39" src="https://user-images.githubusercontent.com/304423/72174259-8f1a3280-33d9-11ea-9c66-a6b6ec276cfa.png">

With basic averaging (not exponential):

<img width="1042" alt="Skärmavbild 2020-01-10 kl  21 06 41" src="https://user-images.githubusercontent.com/304423/72183065-30f74a80-33ed-11ea-8de8-188d04b12e5a.png">

A newer screenshot (lots of 1PCT pools at the top now):

<img width="1054" alt="Skärmavbild 2020-01-13 kl  15 47 52" src="https://user-images.githubusercontent.com/304423/72279476-82464a80-3636-11ea-8ea9-2a9150ac4166.png">

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Implement Cardano.Pool.Ranking

With non-myopic desriability as specified by the spec.

Non-myopic member rewards are not implemented yet.

jonathanknowles commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -144,6 +146,7 @@ data StakePool = StakePool
     , stake :: Quantity "lovelace" Word64
     , production :: Quantity "block" Word64
     , apparentPerformance :: Double
+    , desirabilitiy :: Double
```
```suggestion
    , desirability :: Double
```

Anviking commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -352,12 +355,44 @@ newStakePoolLayer tr db@DBLayer{..} nl metadataDir = StakePoolLayer
+    jormungandrEnv :: P.Env
+    jormungandrEnv = P.Env
+        { P.leaderStakeInfluence = 0
+        , P.punishmentHarshness = 1 -- TODO: What should this be?
```
Need to decide on this.

<img width="756" alt="Skärmavbild 2020-01-13 kl  15 22 04" src="https://user-images.githubusercontent.com/304423/72263123-85324280-3618-11ea-9728-a4e2720d0f9a.png">

Anviking replied:
Don't think it is there at all in the latest version of the spec. And not needed in this PR anyway. Removed.

Anviking commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -352,12 +355,44 @@ newStakePoolLayer tr db@DBLayer{..} nl metadataDir = StakePoolLayer
+        { P.leaderStakeInfluence = 0
+        , P.punishmentHarshness = 1 -- TODO: What should this be?
+        , P.optimalNumberOfPools = 100
+        , P.totalRewards = 1000000*1000000 -- TODO: What should this be?
```
Will determine the influence of fixed `cost` on the ranking.

Seems non-trivial (but perhaps possible) to calculate from jörmungandr's reward constants. If e.g. tx fees can affect it significantly, it might be tricky to take into account.

I suspect it's okay-ish to ship with a hard-coded estimation. But pools with high `cost` will be ranked unfairly.

Anviking replied:
*will try calculating from jörmungandr's config params*

https://github.com/input-output-hk/chain-libs/blob/master/chain-impl-mockchain/src/rewards.rs#L147-L190 🤔

Anviking replied:
Ideally we wouldn't duplicate the calculation. Hopefully we can ask the Haskell Shelley Node for `R` over the LocalStateQuery protocol, when we get to it.

Anviking replied:
Not needed in this PR. Will be tackled in #1276

KtorZ commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,278 @@
+withRewards :: Rewards -> (Double -> Double) -> Rewards
+withRewards (Rewards r) f = Rewards $ f r
+
+type Lovelace = Double
```
Let's not use type aliases for immediate obvious types. Especially with double indirections `type Cost = Lovelace = Double`. 
Type aliases are harming readability more than they're helping. Prefer:

- newtypes if there are properties that need to hold for that type
- record-types to disambiguate arguments passed to functions. e.g.

```hs
poolMemberReward
    :: Rewards -- \hat{f}
    -> Cost -- c
    -> Margin  -- m
    -> RelativeStakeOf "member" -- t
    -> RelativeStakeOf "pool" -- \sigma
    -> Double
poolMemberReward
```

becomes:

```hs
poolMemberReward 
    :: PoolParameters
    -> RelativeStakeOf "member"
    -> RelativeStakeOf "pool"
    -> Double
poolMemberReward PoolParameters{cost,margin,rewards} t σ
```

Having said that, in this context, I'd opt for sticking to the paper's terminology as much as possible to make review and comparison easier. Cf other comment about identifiers.

Anviking replied:
>  I'd opt for sticking to the paper's terminology

That was my original intent. 

But I later found myself mixing up the identifiers and making severe mistakes. That's why I started introducing longer names.

It should now be a nice mixture of both, with the records having descriptive names, but the formulas using let-bindings to the short names matching the spec.

KtorZ commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,278 @@
+-- Types
+--------------------------------------------------------------------------------
+
+data Env = Env
```
Avoid generic names like `Env` or `Config` which gives poor information about what the type really is about (unless meant for a qualified usage). Prefer something like `IncentiveParameters`

KtorZ commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,278 @@
+    , cost :: Lovelace
+    , margin :: Margin
+    , recentAvgPerformance :: Double
+      -- ^ An already averaged performance-value. This should be an exponential
```
~~should be~~ could be. Be careful, these words have a precise meaning.

Anviking replied:
I was looking at the old spec -> explains it.

KtorZ commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,278 @@
+-- (Kant et al, 2019)
+
+-- @Rank@ and desirability here are temporary concepts that should not be
+-- confused with the ordering of stake-pools in Daedalus.
```
I do not understand this comment, can you clarify :thinking:  ?

Anviking replied:
Under the constraint "the wallet should order pools by non-myopic member rewards" this holds by definition. Under the constraint "the wallet should order pools by desirability" — no.

KtorZ commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,278 @@
+
+-- | How desirable a pool would be if it were saturated.
+--
+-- (6.6.1 Pool Desirability and Ranking)
```
*5.6.1

KtorZ commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,278 @@
+-- | How desirable a pool would be if it were saturated.
+--
+-- (6.6.1 Pool Desirability and Ranking)
+saturatedDesirability
```
There's no such thing as "unsaturatedDesirability" so what about calling this "desirability" to better match the spec?

KtorZ commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,278 @@
+saturatedDesirability env cost' margin' (RelativeStake leaderStake') =
+        if saturatedPoolRewards > cost'
+        then (saturatedPoolRewards - cost') * (1 - margin')
+        else 0
```
Alternatively, 

```hs
    | saturedPoolRewards <= cost' = 0
    | otherwise                   = (saturatedPoolRewards - cost') * (1 - margin')
```

to be closer to the spec notation + guards read better than if-then-else.

KtorZ commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,278 @@
+    (RelativeStake totPoolStake) =
+    if rew > cost'
+    then (rew - cost') * (1 - margin') * (memberStake / totPoolStake)
+    else 0
```
The calculation of pool and member rewards is way out of scope for this U/S. The calculation of `R` shouldn't force us to re-implement the whole reward mechanism. Rather, this is something we want to fetch from the node's.

KtorZ commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,278 @@
+-- \hat{f} -> pool rewards (includes average!)
+-- c -> pool costs in ADA
+-- m : [0, 1]  -> margin
+-- σ : [0,1]
```
What about using these identifiers consistently below instead of ad-hoc names like `saturedPoolRewards` ? Also, after a quick look, it looks like we truly need only a subset of these which we can categorize as follows:

```
Epoch Parameters
================

identifier   | domain / definition      | description
---          | ---                      | ---
R            | [0, ∞)                   | the total available rewards for the epoch (in ada).
a0           | [0, ∞)                   | owner-stake influence on pool rewards.
k            | [[0, ∞)                  | desired number of pools
z0           | 1/k                      | size of a saturated pool


Pool's Parameters
=================

identifier | domain / definition | description
---        | ---                 | ---
c          | [0, ∞)              | pool's costs
f          | [0, ∞)              | pool's rewards
m          | [0, 1]              | pool's margin
p_apparent | [0, 1]              | pool's apparent performance
s          | [0, ∞)              | relative stake of pool leader (i.e. pledge amount / total pool's stake)
σ          | [0, 1]              | relative stake of the pool, the fraction of the absolute stake that the pool controls
```

Anviking replied:
Put it in actual haddock table syntax:

<img width="452" alt="Skärmavbild 2020-01-16 kl  18 07 43" src="https://user-images.githubusercontent.com/304423/72551758-ce42fa80-3895-11ea-884b-c686937f0f5a.png">

I realised one could argue this table to be superfluous as we have these definitions (mostly I believe) in the code itself, but might be nice to have 🤷‍♂

KtorZ commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,278 @@
+    -> Rank
+    -> RelativeStakeOf "pool"
+    -> RelativeStakeOf "pool"
+nonMyopicStakeOfPool env _leaderStake' rank poolStake
```
This is out-of-scope.

KtorZ commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,278 @@
+    -> RelativeStakeOf "member"
+    -> [pool]
+    -> [(pool, Double)]
+nonMyopicMemberRewards env convert memberStake pools =
```
This is out-of-scope.

KtorZ reviewed (changes requested):
I am a bit confused. There seems to be a lot of focus on the calculation of rewards, rewards per stakeholder and  whereas.. the point of the user story was to compute the `desirability` in order to rank pools.

This boils down to:

```hs
desirability (R, z0, a0) c m s p_apparent
    | f_satured <= c = 0
    | otherwise      = (f_satured - c) * (1 - m)
  where
    f_satured = p_apparent * optimalReward (R, z0, a0) z0 s

optimalReward (R, z0, a0) σ s = 
    R / (1 + a0) * (σ' + s'*a0*(σ' - s'*(z0-σ')/z0) / z0)
  where
    σ' = min σ z0              
    s' = min s z0
```

where `p_apparent` is already computed by the `Metrics` module. There's the question of `R` which we are definitely not looking to re-compute (unless trivial) but, fetch from the node. 

Please revise and focus on the calculation of `desirability` as highlighted above. I'd recommend sticking with the same terminology as in the specs to make the mapping immediate and avoid potential translation errors. Keep types simple (`Double` are fine here, really). Leave the calculation or fetching of `R` to another PR.

Anviking commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,158 @@
+-- Types
+--------------------------------------------------------------------------------
+
+newtype a >= (b :: Nat) = Bounded { unBounded :: a }
```
Will likely remove.

KtorZ replied:
What about using something more like `newtype NonNegative a = NonNegative a` ? A phantom type sounds a bit convoluted.

Anviking replied:
Tbh, I really like how `Double >= 0` reads.  But, yes, that seems decent. There are parse issues with `>= 0` and hlint.

So going with 2:

### 1. `Double >= 0`, `Double >= 1`
- Tbh, I do really like how this reads
- TypeOperators, Nats
- Looks like we only need the `>= 0` and `>= 1` cases, so it is more general than we need.
- Causes parse-issues with both stylish-haskell and hlint. Even with `hlint` built on `master`.

### 2. `NonNegative Double`, `Positive Double`
- Similar to 1, but less fancy

### 3. Enforce in each newtype: `Lovelace`, `TotalRewards`, ...
- We'd need *a lot of* new types. Seems bad to me.
- `TotalRewards` are in fact in Lovelace, but would have to be separate, unrelated newtypes.

Anviking commented on lib/core/test/unit/Cardano/Pool/PerformanceSpec.hs:
```diff
@@ -0,0 +1,131 @@
+
+-- | If two pools differ only in one field, and one is more desirable than the
+-- other, how should the field values compare?
+allElseEqualProperty
```
Sadly the important `(<)` `(>)` distinction on the call-sites ended up less readable than I'd hoped.

Anviking replied:
Think I can simplify the structure a bit. Will do tomorrow.

Anviking commented on lib/core/src/Cardano/Pool/Performance.hs:
```diff
@@ -0,0 +1,155 @@
+    saturatedPoolRewards :: Double
+    saturatedPoolRewards =
+        let
+            a0 = unBounded $ leaderStakeInfluence env
```
I am still using the `env` variable name after renaming the type. Should change.

I'm still not sure of the name of this module. But I like importing it as `qualified`. It might affect the name of what is now `RankingConstants`.

KtorZ commented on lib/core/src/Cardano/Pool/Ranking.hs:
```diff
@@ -0,0 +1,170 @@
+        -- ^ technically \hat{p} in the spec
+    in
+        (p * _R) / (1 + a0)
+        * (z0 + ((min s z0) * a0))
```
May I suggest to define the "optimalRewards" at the top-level as:

```hs
optimalReward (R, z0, a0) σ s = 
    R / (1 + a0) * (σ' + s'*a0*(σ' - s'*(z0-σ')/z0) / z0)
  where
    σ' = min σ z0              
    s' = min s z0
```

Where the saturated pool reward can be obtained by replacing `σ` by `z0`:

```hs
f_satured = p_apparent * optimalRewards (R, z0, a0) z0 s
```

Anviking replied:
Sounds good, but maybe not necessary now. Causes minor cascading changes.

Anviking commented on lib/core/src/Cardano/Pool/Ranking.hs:
```diff
@@ -0,0 +1,188 @@
+      -- ^ \gamma
+    , optimalNumberOfPools :: Int >= 1
+      -- ^ k
+    , totalRewards :: Lovelace >= 1
```
Hm… not sure why I forced R >= 1 instead if 0 🤔

Anviking commented on lib/core/test/unit/Cardano/Pool/RankingSpec.hs:
```diff
@@ -0,0 +1,226 @@
+                (property prop_saturatedPoolRewardsNonNegative)
+
+        describe "mkRelativeStake" $ do
+            it "stake > R, R >= 1 => mkRelativeStake ∈ [0,1] (even with conversions)"
```
I was afraid of rounding errors (not sure if rightfully so).

KtorZ commented on lib/core/src/Cardano/Pool/Ranking.hs:
```diff
@@ -0,0 +1,206 @@
+-- | Determines z0, i.e 1 / k
+saturatedPoolSize :: EpochConstants -> RelativeStakeOf "pool"
+saturatedPoolSize constants =
+    RelativeStake $ 1 / fromIntegral (getPositive $ optimalNumberOfPools constants)
```
Why is this of type `RelativeStakeOf` ? I thought the point of these newtypes was to provide insights on the semantic of the various numbers, `σ` stands for the relative stake of the pool and, although they have the same dimension, labelling both as "RelativeStakeOf" feels misleading. 

Aype that captures what the domain actually is would be preferable (which is what types for in the first place, representing a range of possible values): `Ratio 1 0 ` or `Percentage` ?

Anviking replied:
I introduced `RelativeStakeOf` to protect against mixing up `\sigma` and `s`.

Although, at that time, most functions took only the pool parameters they needed. And it was easy to get the order wrong. Now both `desirability` and `saturatedPoolSize` take `Pool`, and the `RelativeStakeOf tag` tag might be less useful.

KtorZ commented on lib/core/src/Cardano/Pool/Ranking.hs:
```diff
@@ -0,0 +1,206 @@
+    { leaderStakeInfluence :: NonNegative Double
+      -- ^ a_0
+    , optimalNumberOfPools :: Positive Int
+      -- ^ k
```
`k` is referred to as the _"desired number of pools"_, which is different from an _"optimal"_ number. Rather, we chose to fix `k` and then, determine what other parameters should be. But, chosing `k=10` or `k=100` or `k=1000` is rather arbitrary and simply witnesses of how decentralized we want the network to be.

KtorZ commented on lib/core/src/Cardano/Pool/Ranking.hs:
```diff
@@ -0,0 +1,206 @@
+    } deriving (Show, Eq, Generic)
+
+data Pool = Pool
+    { leaderStake :: RelativeStakeOf "pool leader"
```
A comment to link this to `s` is missing. Also, same remark about the choice of type regarding "RelativeStakeOf" applies here too.

```hs
-- s 
```

KtorZ commented on lib/core/src/Cardano/Pool/Ranking.hs:
```diff
@@ -0,0 +1,206 @@
+data Pool = Pool
+    { leaderStake :: RelativeStakeOf "pool leader"
+    , cost :: Lovelace
+    , margin :: Margin
```
`margin` and `leaderStake` have so-to-speak the same dimension, so arguably they should have the same type.

KtorZ commented on lib/core/src/Cardano/Pool/Ranking.hs:
```diff
@@ -0,0 +1,206 @@
+      -- ^ An already averaged performance-value
+    } deriving (Show, Eq, Generic)
+
+newtype Lovelace = Lovelace { getLovelace :: NonNegative Double }
```
Why are Lovelace represented as `Double` :thinking: ? These clearly are non-negative integral values?  Also, the extra `Lovelace` wrapping seems superfluous since we don't attach any properties to the fact that they are Lovelace apart from:

- The fact that they are non-negative
- The fact that they are integers

Which can be entirely captured with "NonNegative Integer", or, an existing type: "Natural".

KtorZ commented on lib/core/src/Cardano/Pool/Ranking.hs:
```diff
@@ -0,0 +1,206 @@
+unsafeToPositive :: (Ord a, Show a, Num a) => a -> (Positive a)
+unsafeToPositive x
+    | x >= 0    = Positive x
+    | otherwise = error $ "unsafeToPositive: " ++ show x ++ " > 0 does not hold"
```
Note that `Positive a` in this module is only used for `Positive Int` which can be substituted with `Natural` from the `base` package. The `Positive` data type is therefore redundant.

Anviking replied:
Sorry, I got the condition wrong here. It should be `x > 0`, which is not captured by `Natural`.

This is to prevent division-by-zero if anyone sets `k = 0`.

I should add tests for these `unsafeTo`-constructors if we intend to have them.

KtorZ commented on lib/core/src/Cardano/Pool/Ranking.hs:
```diff
@@ -0,0 +1,206 @@
+    { leaderStake :: RelativeStakeOf "pool leader"
+    , cost :: Lovelace
+    , margin :: Margin
+    , recentAvgPerformance :: NonNegative Double
```
`NonNegative Double` would suppose that the performance could be greater than 1, which isn't the case. Incidentally `margin`, `leaderStake`  and `recentAvgPerformance` all have the same unit / type which should be equivalent to a fractional number between 0 and 1.

Anviking replied:
>  performance could be greater than 1, which isn't the case

I cannot conclude this from the spec.

> Even when a pool’s true performance is 1, its actual rewards might be less than its optimal rewards, because of the randomness in re. Likewise, sometimes a pool’s actual rewards will be **higher than their optimal rewards**, if they are **lucky** in the leader election process.
>
> **These effects will balance each other**, so that a well performing pool will get their optimal rewards on average over multiple epochs.

and from the formulas I don't see how `\bar{p}` would be constrained to `[0,1]`:

<img width="801" alt="Skärmavbild 2020-01-17 kl  11 48 04" src="https://user-images.githubusercontent.com/304423/72606639-3d186600-391f-11ea-834d-2ca2c5d97213.png">

I have another concern about clamping the **pre**-averaged performances, especially, with `min 1` , but that is separate to this.

Anviking replied:
If there is a choice to be made whether to restrict to `[0,1]` or not it seems better to me to not make that choice here. Especially if the type of `recentAvgPerformance` is going to error when a value is out of bounds.

KtorZ replied:
Fair enough :+1:

KtorZ reviewed (changes requested):
If using dedicated data-types for the various fields we should make sure to use relevant types to capture the right domain of values. 
I'll have a look at the properties and will comment on this bit in a moment.

KtorZ committed:
simplifies allElseEqualProperty + apply coding standards

- Favor explicit importsover qualified imports
- Avoid duplication in function signatures and function body
- Use type-signatures for intermediate functions
- Wrap lines that are too long

Anviking committed:
Add unsafeTo-tests and fixups

KtorZ commented on lib/core/test/unit/Cardano/Pool/RankingSpec.hs:
```diff
@@ -0,0 +1,254 @@
+    forAll arbitrary $ \v1 v2 -> do
+        let (higherStake, lowerStake) = sortTuple (v1, v2)
+        let z0 = unRelativeStake $ saturatedPoolSize constants
+        higherStake > z0 ==>
```
Side note but, this pre-condition should really be `lowerStake > z0`, shouldn't it?

Anviking replied:
If `lowerStake > z0` and `higherStake >= lowerStake`, then both pools would be over-saturated.

It's enough to have the *one* pool be oversaturated, for it to be "more oversaturated" than the other one.

(So I guess I'm defining `oversaturation = max 0 (sigma - z0)`)

KtorZ replied:
But.. the `stake` of the pool doesn't even enter into account in the computation of the desirability. So, the property reduces to: `forall d. d === d ~ True`. This property is a tautology. Let's remove it and feel free to drop a comment in the source module about the effect of over-saturation.

Anviking replied:
> But.. the stake of the pool doesn't even enter into account in the computation of the desirability. So, the property reduces to: forall d. d === d ~ True. This property is a tautology. Let's remove it and feel free to drop a comment in the source module about the effect of over-saturation.

I think my response above applies here. Yes it is a tautology, but the moment we add a `stake` field (as you indirectly suggested in a previous comment, or when we implement non-myopic member rewards), it will not longer be.

Anviking replied:
Although, if you insist, let me know…

There is already a comment on `desirability`.

KtorZ replied:
> the moment we add a stake field (as you indirectly suggested in a previous comment, or when we implement non-myopic member rewards), it will not longer be.

Then, we'll add the property at this moment :+1: 
I insist :)

KtorZ commented on lib/core/test/unit/Cardano/Pool/RankingSpec.hs:
```diff
@@ -0,0 +1,254 @@
+    :: Pool
+    -> EpochConstants
+    -> Property
+prop_oversaturationDoesNotInfluenceDesirability pool constants =
```
I am not entirely convinced by what this property tests. Of course, the pool's stake doesn't influence the desirability because it's not even a parameter of the equation!

Anviking replied:
I think it's easy to guess that `desirability` should take oversaturation into account. I wanted to make it obvious that it doesn't.

If we ever add `stake` as a field to `Pool` it will not be obvious from the type-signatures alone.

KtorZ commented on lib/core/test/unit/Cardano/Pool/RankingSpec.hs:
```diff
@@ -0,0 +1,254 @@
+    -> EpochConstants
+       -- ^ Constants used
+    -> Property
+allElseEqualProperty expectedEffect pool constants =
```
Good one :+1:

Anviking replied:
```Haskell
        , HasField' field Pool a
        , KnownSymbol field
```

Nice simplification 👍 , suspected that was theoretically possible

KtorZ committed:
disable buggy generic shrinkers on ranking specs

KtorZ commented on lib/core/test/unit/Cardano/Pool/RankingSpec.hs:
```diff
@@ -0,0 +1,254 @@
+    shrink = map unsafeMkMargin
+        . filter (\a -> a >= 0 && a <= 1)
+        . map getMargin
+        . shrink
```
The arbitrary instances for `Margin` and `RelativeStakeOf` are identical which re-enforce the argument that these should really be of the same type.

KtorZ commented on lib/core/test/unit/Cardano/Pool/RankingSpec.hs:
```diff
@@ -0,0 +1,254 @@
+                $ property prop_saturatedPoolRewardsReduces
+
+            it "saturatedPoolRewards >= 0"
+                $ property prop_saturatedPoolRewardsNonNegative
```
:+1:

Anviking committed:
Rename Margin -> Ratio

Anviking committed:
Remove RelativeStakeOf for Ratio

Anviking committed:
Lovelace can be a Word64 as would be more appropriate

Anviking committed:
adjust unsafeMkRelativeStakeBounds

KtorZ commented on lib/core/src/Cardano/Pool/Ranking.hs:
```diff
@@ -0,0 +1,198 @@
+      -- Should mostly be in the range [0, 1], but lucky pools may exceed 1.
+    } deriving (Show, Eq, Generic)
+
+newtype Lovelace = Lovelace { getLovelace :: Word64 }
```
Minor note: I find "Natural" conceptually better than `Word64` (in this context at least, but generally speaking) to describe unbounded positive integers. `Word64` is something we would find at boundaries of the system like databases or types in network serializations. Yet, conceptually, the _right_ type to describe positive integers is `Natural`.

KtorZ commented on lib/core/src/Cardano/Pool/Ranking.hs:
```diff
@@ -0,0 +1,198 @@
+                          ++ "not in range [0, 1]"
+
+unsafeMkRelativeStake :: Lovelace -> EpochConstants -> Ratio
+unsafeMkRelativeStake (Lovelace stake) constants =
```
Minor note: I find it inconsistent to have within the same module names like `unsafeToXXX` vs `unsafeMkXXX` while the constructors fulfill the same goal. Having said that, that's probably my OCD being triggered but there's nothing intrinsically wrong with either of those names.

Anviking replied:
>  I find it inconsistent to have within the same module names like

Yes — I agree, which one do you prefer?

KtorZ replied:
My own preference would be `unsafeMk` but I have no valid rationale to justify this.

Anviking replied:
Done.

Anviking replied:
I just realised that `unsafeMkRelativeStake` is fatally incorrect. It calculates `stake / totalRewards`, and not `stake / totalStake`.

Anviking committed:
Shrink more

Anviking committed:
Make allElseEqualProperty clearer & sort correctly

Anviking committed:
unsafeTo -> unsafeMk

Anviking committed:
fixup: more shrinking

Anviking commented on lib/core/test/unit/Cardano/Pool/RankingSpec.hs:
```diff
@@ -0,0 +1,284 @@
+
+instance Arbitrary Ratio where
+    arbitrary = unsafeMkRatio <$> choose (0, 1)
+    shrink _ = [] -- TODO: Implement a shrinker that works.
```
It was the `Ratio` shrinker that was faulty. I removed it, and re-added `genericShrink` for `Pool`.

Anviking committed:
Remove faulty unsafeMkRelativeStake

It compared the stake with totalRewards instead of totalStake.

total stake is not included in `EpochConstants`. I think it makes sense
to just remove `unsafeMkRelativeStake` and let callers deal with
constructing a correct `Ratio`.

Anviking committed:
Tweak comment

Anviking committed:
Remove prop_oversaturationDoesNotInfluenceDesirability

It is currently a tautology. May be re-added when there is a stake-field
in the Pool record.
